File: csharp_solutions/easy/TwoSum.cs
namespace TwoSum
{
    public class Solution
    {
        public int[] TwoSum(int[] nums, int target)
        {
            var diffMap = new Dictionary<int, int>();
            for (int i = 0; i < nums.Length; i++)
            {
                int difference = target - nums[i];
                if (diffMap.ContainsKey(difference))
                {
                    return new int[] { diffMap[difference], i };
                }
                diffMap.Add(difference, i);
            }
            return new int[] { -1, -1 };
        }
    }
}

File: 02_session/easy/two_sum.rb
# frozen_string_literal: true

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
  diff_map = {}
  nums.each_with_index do |n, i|
    diff = target - n
    return [diff_map[diff], i] if diff_map.key?(diff)

    diff_map[n] = i
  end
end

# 5min

# NOTES
# algo/structure: hashmap, array
# time: O(n)
# space: O(n)

# Use hashmap to cache values and avoid nested looping in O(n^2) time
